src/roswire/ros2/state.py

```diff
@@ -28,31 +35,61 @@ class ROS2SystemState(SystemState):
         A mapping from topics to the names of subscribers to that topic.
     services: Mapping[str, Collection[str]]
         A mapping from services to the names of providers of that service.
+    clients: Mapping[str, Collection[str]]
+        A mapping from services to the names of clients of that service.
+    action_servers: Mapping[str, Collection[str]]
+        A mapping from actions to the names of providesr of that action.
+    action_clients: Mapping[str, Collection[str]]
+        A mapping from actions to the names of clients of that action
     nodes: AbstractSet[str]
         The names of all known nodes running on the system.
     topics: AbstractSet[str]
         The names of all known topics on the system with at least one
         publisher or one subscriber.
+    service_names: AbstractSet[str]
+        The name of all the known services on the system with at least
+        one server or client
+    action_names: AbstractSet[str]
+        The name of all the known actions on the system with at least one
+        one action server or client
     """
 
     publishers: Mapping[str, Collection[str]]
     subscribers: Mapping[str, Collection[str]]
     services: Mapping[str, Collection[str]]
+    service_clients: Mapping[str, Collection[str]]
+    action_servers: Mapping[str, Collection[str]]
+    action_clients: Mapping[str, Collection[str]]
     nodes: AbstractSet[str] = attr.ib(init=False, repr=False)
     topics: AbstractSet[str] = attr.ib(init=False, repr=False)
+    service_names: AbstractSet[str] = attr.ib(init=False, repr=False)
+    action_names: AbstractSet[str] = attr.ib(init=False, repr=False)
 
     def __attrs_post_init__(self) -> None:
         nodes: Set[str] = set()
         nodes = nodes.union(*self.publishers.values())
         nodes = nodes.union(*self.subscribers.values())
         nodes = nodes.union(*self.services.values())
+        nodes = nodes.union(*self.service_clients.values())
+        nodes = nodes.union(*self.action_servers.values())
+        nodes = nodes.union(*self.action_clients.values())
 
         topics: Set[str] = set()
         topics = topics.union(self.publishers)
         topics = topics.union(self.subscribers)
 
+        service_names: Set[str] = set()
+        service_names = service_names.union(self.services)
```

Comment:
This might produce unexpected results. `self.services` has type `Mapping[str, Collection[str]]`, but `union` is expecting `Iterable[str]`.